dash/util/s3_migration.py

```
@@ -0,0 +1,361 @@
+
+    Returns number of objects downloaded.
+    """
+    transfer = S3Transfer(
+        client,
+        config=TransferConfig(
+            max_concurrency=max_concurrency, multipart_threshold=16 * 1024 * 1024
+        ),
+    )
+    objs = list_objects(client, bucket, prefix)
+
+    if not objs:
+        logging.info("No objects under s3://%s/%s", bucket, prefix)
+        return 0
+
+    logging.info(
+        "Planning to download %d objects (%.2f MiB) from s3://%s/%s to %s",
+        len(objs),
+        sum(obj.size for obj in objs) / 1024 / 1024,
+        bucket,
+        prefix,
+        dest_dir,
+    )
+
+    dest_dir.mkdir(parents=True, exist_ok=True)
+
+    if dry_run:
+        for obj in tqdm(objs, unit="file"):
+            logging.info(
+                "DRY-RUN download -> %s", dest_dir / relative_key_path(obj.key, prefix)
+            )
+        return 0
+
+    def _download(obj: S3Object):
+        f_downloaded = 0
+        rel = relative_key_path(obj.key, prefix)
+        target = dest_dir / rel
+        target.parent.mkdir(parents=True, exist_ok=True)
+        try:
+            if target.exists() and target.stat().st_size == obj.size:
+                pass
+            else:
+                transfer.download_file(bucket, obj.key, str(target))
+            f_downloaded += 1
+        except Exception as e:
+            logging.error(
+                "Failed to download s3://%s/%s -> %s: %s", bucket, obj.key, target, e
+            )
+            raise
+        return f_downloaded
+
+    downloaded = 0
+
+    with concurrent.futures.ThreadPoolExecutor(max_workers=max_concurrency) as pool:
+        futures = [pool.submit(_download, obj) for obj in objs]
+        for f in tqdm(
+            concurrent.futures.as_completed(futures), total=len(futures), unit="file"
+        ):
+            try:
+                f_downloaded = f.result()
+                downloaded += f_downloaded
+            except Exception:
```

Could we raise this exception to help with debugging?